drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -1442,787 +1571,1118 @@ void RasterizerCanvasGLES2::canvas_render_items(Item *p_item_list, int p_z, cons
-					}
+		//				glActiveTexture(GL_TEXTURE0);
+		//				_canvas_item_render_commands(ci, NULL, reclip, material_ptr); //redraw using light
 
-				} break;
-				case RasterizerStorageGLES2::Shader::CanvasItem::BLEND_MODE_SUB: {
+		//				state.using_light = NULL;
+		//			}
 
-					glBlendEquation(GL_FUNC_REVERSE_SUBTRACT);
-					if (storage->frame.current_rt && storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_TRANSPARENT]) {
-						glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE, GL_SRC_ALPHA, GL_ONE);
-					} else {
-						glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE, GL_ZERO, GL_ONE);
-					}
-				} break;
-				case RasterizerStorageGLES2::Shader::CanvasItem::BLEND_MODE_MUL: {
-					glBlendEquation(GL_FUNC_ADD);
-					if (storage->frame.current_rt && storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_TRANSPARENT]) {
-						glBlendFuncSeparate(GL_DST_COLOR, GL_ZERO, GL_DST_ALPHA, GL_ZERO);
-					} else {
-						glBlendFuncSeparate(GL_DST_COLOR, GL_ZERO, GL_ZERO, GL_ONE);
-					}
+		//			light = light->next_ptr;
+		//		}
+
+		//		if (light_used) {
+
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::USE_LIGHTING, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::USE_SHADOWS, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_NEAREST, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF3, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF5, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF7, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF9, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF13, false);
+
+		//			state.canvas_shader.bind();
+
+		//			ris.last_blend_mode = -1;
+		//		}
+	}
+
+	if (reclip) {
+		join = false;
+	}
+
+	// non rects will break the batching anyway, we don't want to record item changes, detect this
+	if (_detect_batch_break(ci)) {
+		join = false;
+		r_batch_break = true;
+	}
+
+	return join;
+}
+
+bool RasterizerCanvasGLES2::_detect_batch_break(Item *ci) {
+	int command_count = ci->commands.size();
+
+	// It is hard to know what this number should be, empirically,
+	// and this has not been fully investigated. It works to join single sprite items when set to 1 or above.
+	// Note that there is a cost to increasing this because it has to look in advance through
+	// the commands.
+	// On the other hand joining items where possible will usually be better up to a certain
+	// number where the cost of software transform is higher than separate drawcalls with hardware
+	// transform.
+
+	// if there are more than this number of commands in the item, we
+	// don't allow joining (separate state changes, and hardware transform)
+	// This is set to quite a conservative (low) number until investigated properly.
+	const int MAX_JOIN_ITEM_COMMANDS = 16;
```

Comment:
I'm not sure I understand this properly. Is this to ensure that an object with many commands (like a tilemap) is assigned its own batch and avoids doing all the software transforms associated with batching?

Reply:
Yes, pretty much. Essentially the problem is this: Any item that contains commands that are default (i.e. not handled by software transform and the batching renderer) should not be joined.

In order to work this out, it does a lookahead through the commands, which could potentially be very expensive. As such it makes sense to put a limit on this to some small number, which will catch nearly all cases which need joining, but not be overly expensive in the case of items with large numbers of commands.

There may be a far more efficient way of handling this. Ideally it would be nice in general to mark some extra data like this on the items themselves, and some dirty flags for when things change in the item. There's a lot of inefficiencies throughout to deal with the historical design.

In practice I don't think this will be a problem in 3.2 to do like this, but we could look at some better approaches in 4.0 where it might be possible to move some of this processing further up into the visual server.

Reply:
Right, that makes sense. Thank you. 

We should consider making this a project setting as well.

Reply:
Can do, especially for testing, if you think this a good idea. :+1: 

I do suspect that once we have figured out a good value, it won't need to be user editable .. I'm conscious there's a problem with 'project settings creep' at the moment, so I'd hesitate to go overboard on this. Plus it won't be at all obvious to end users what this does. Maybe the project settings overchoice problem will be addressed in 4.0 with 'advanced settings'.

Another candidate for the project settings is the max vertex buffer size, which at the moment is hard coded in initialize().

Reply:
Ah, yes, please make max vertex size a project setting as well. I meant to comment on that but forgot.